### PR TITLE
Validate member types

### LINKED ISSUE #34864: [API Compat] Changing a member return type should be detected as a breaking change
### Describe the bug
Changing the return type of a member, whether its a field, property or method isn't detected by the API Compat tool.

### To Reproduce
The following zip files has two assemblies, `before.dll` and `after.dll`, each with a single class `C` having three members. The member names are the same for the two assemblies, but their return type is different. These are breaking changes, and will result in `MissingFieldException`/`MissingMemberException`/`MissingMethodException` at runtime.

File: [diff.zip](https://github.com/dotnet/sdk/files/12418773/diff.zip)

Before.dll:
```csharp
public class C
{
    public readonly string Field = "f";
    public int Property => 123;
    public double Method() => 45.6;
}
```

After.dll:
```csharp
public class C
{
    public readonly int Field = 0;
    public string Property => "p";
    public DateTime Method() => DateTime.Now;
}
```

Run
```
apicompat -l=./before.dll -r=./after.dll
```

Expected: APICompat reports three breaking changes.
Actual: "APICompat ran successfully without finding any breaking changes."

### Further technical details
Reproduced with both:
 - Latest release: 7.0.400+8e1f76c1f41ecd37935288c617c53342533db704
 - Latest nightly build: 9.0.100-alpha.1.23417.7+4249701de6f005c670c7f371141e772d02f89156

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/sdk/issues/34864

cc @MrJul

### CODE REVIEW

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs [hunk 1]
```diff
@@ -531,5 +531,56 @@ public static class First
...
+
+            IEnumerable<CompatDifference> differences = differ.GetDifferences(left, right);
+
+            Assert.Equal(new[]
+            {
+                CompatDifference.CreateWithDefaultMetadata(DiagnosticIds.MemberMustExist, string.Empty, DifferenceType.Removed, "F:CompatTests.First.S"),
+                // CompatTests.First.Prop.set isn't reported as the return types match: 'void'.
+                CompatDifference.CreateWithDefaultMetadata(DiagnosticIds.MemberMustExist, string.Empty, DifferenceType.Removed, "M:CompatTests.First.get_Prop"),
+                CompatDifference.CreateWithDefaultMetadata(DiagnosticIds.MemberMustExist, string.Empty, DifferenceType.Removed, "M:CompatTests.First.M")
+                // CompatTests.First.E_add and CompatTests.First.E_remove aren't reported as the symbol's DisplayString doesn't include the parameter type.
```
Comment (ViktorHofer) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:582]: Reached out to @CyrusNajmabadi offline to clarify if this behavior is expected.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs [hunk 2]
```diff
@@ -531,5 +531,56 @@ public static class First
...
+
+namespace CompatTests
+{
+  public class First
+  {
+    public string S;
+    public bool Prop { get; set; }
+    public string M() => null;
+    public delegate void SampleEventHandler(object sender, EventArgs e);
+    public event SampleEventHandler E;
```
Comment (carlossanlop) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:549]: What about nested member types? Class, struct, enum?

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs (hunk 1 above)
Comment (carlossanlop) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:582]: What is "DisplayString" here?

**Review by carlossanlop — APPROVED** [review]
Left two questions, but lgtm.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:582]: The `DisplayString` is the string that is returned from the ISymbol's `GetDisplayString()` method. It is configurable but it currently doesn't include the parameter type or the parameter at all for the event method.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:549]: This is just about members: fields, properties, events and methods. They behave the same in nested classes/structs.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:582]: discussed offline.  i feel like this is expected.  note: this space makes me *very* wary.  It's very much *not* an intended use case of .ToDisplayString (which, as the name suggests, is entirely designed around just displaying symbols to a user, not for any sort of comparison/versioning story).

Comment (carlossanlop) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:582]: @CyrusNajmabadi should comparisons be done between DocIds instead?

**Pull request merged**